src/ccall.cpp

```diff
@@ -1648,166 +1728,242 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
+            v = julia_to_address(largty, jargty_in_env, unionall_env, arg,
+                                 ai + 1, ctx, &needStackRestore);
+            assert((!toboxed && !byRef) || isa<UndefValue>(v));
         }
 
-        Value *v = julia_to_native(largty, toboxed, jargty, arg, addressOf, byRef,
-                                   false, ai + 1, ctx, &needStackRestore);
-        bool issigned = jl_signed_type && jl_subtype(jargty, (jl_value_t*)jl_signed_type);
-        argvals[ai + sret] = llvm_type_rewrite(v, largty,
-                ai + sret < fargt_sig.size() ? fargt_sig.at(ai + sret) : fargt_vasig,
-                false, byRef, issigned, ctx);
+        if (isa<UndefValue>(v)) {
+            JL_GC_POP();
+            return jl_cgval_t();
+        }
+        assert(v->getType() == pargty);
+        argvals[ai + sret] = v;
     }
 
+    Value *result = NULL;
+    // First, if the ABI requires us to provide the space for the return
+    // argument, allocate the box and store that as the first argument type
+    bool sretboxed = false;
+    if (sret) {
+        jl_cgval_t sret_val = emit_new_struct(rt, 1, NULL, ctx); // TODO: is it valid to be creating an incomplete type this way?
+        assert(sret_val.typ != NULL && "Type was not concrete");
+        if (!sret_val.ispointer()) {
+            Value *mem = emit_static_alloca(lrt, ctx);
+            builder.CreateStore(sret_val.V, mem);
+            result = mem;
+        }
+        else {
+            // XXX: result needs a GC root here if result->getType() == T_pjlvalue
+            result = sret_val.V;
+        }
+        argvals[0] = emit_bitcast(result, fargt_sig.at(0));
+        sretboxed = sret_val.isboxed;
+    }
+
+    Instruction *stacksave = NULL;
+    if (needStackRestore) {
+        stacksave = CallInst::Create(Intrinsic::getDeclaration(jl_Module,
+                                                               Intrinsic::stacksave));
+        if (savespot) {
+#if JL_LLVM_VERSION >= 30800
+            instList.insertAfter(savespot->getIterator(), stacksave);
+#else
+            instList.insertAfter(savespot, stacksave);
+#endif
+        }
+        else {
+            instList.push_front(stacksave);
+        }
+    }
 
     // make LLVM function object for the target
     // keep this close to the function call, so that the compiler can
     // optimize the global pointer load in the common case
     Value *llvmf;
-    FunctionType *functype = FunctionType::get(sret ? T_void : prt, fargt_sig, isVa);
-
-    if (jl_ptr != NULL) {
-        null_pointer_check(jl_ptr,ctx);
-        Type *funcptype = PointerType::get(functype,0);
-        llvmf = builder.CreateIntToPtr(jl_ptr, funcptype);
+    if (llvmcall) {
+        if (symarg.jl_ptr != NULL) {
+            jl_error("llvmcall doesn't support dynamic pointers");
+        }
+        else if (symarg.fptr != NULL) {
+            jl_error("llvmcall doesn't support static pointers");
```

Comment:
yeah, I know. It's also seems to be fundamentally incompatible with the way we manage the rest of the code pipeline, so it's low on my priority lists of things that need to work. I've written up a replacement plan at https://github.com/vtjnash/julep/wiki/0002:-Enhanced-llvmcall (which I believe should also cover Tim's case below), but waiting for someone to actually care about the implementation. Maybe that person could be you :) :) ;) :)